frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,321 @@
+            return self:isAvailable()
+        end,
+        enabled_func = function()
+            return self:_enabled()
+        end,
+        separator = true,
+    }
+end
+
+
+-- add a button for dictquicklookup
+function ReaderUserHyph:addButton(buttons, pos, word, parentWidget)
+    if word:find("[ ,;-%.\n]") then return end
+
+    if G_reader_settings:readSetting("hyph_user_dict") then
+        table.insert(buttons, pos, {
+            {
+                id = "add_user_dict_entry",
+                text = T(_("Modify hyphenation of \"%1\"."), word),
+                callback = function()
+                    if parentWidget then
+                        UIManager:close(parentWidget)
+                    end
+                    self:modifyUserEntry(word)
+                end,
+            },
+        })
+    end
+end
+
+-------------------------------------------
+-- Helper functions for dictionary entries
+-------------------------------------------
+
+-- checks if suggestion is well formated
+function ReaderUserHyph:checkHyphenation(suggestion, word)
+    if suggestion:find("%-%-") then
+        return false -- two or more consecutive '-'
+    end
+
+    suggestion = suggestion:gsub("-","")
+    if self.ui.document:getLowercasedWord(suggestion) == self.ui.document:getLowercasedWord(word) then
+        return true -- characters match (case insensitive)
+    end
+    return false
+end
+
+-------------------------------------
+-- dictionary, file functions use:
+--    self.dict_file
+--    self.new_dict_file
+--    self.dict
+-------------------------------------
+
+-- opens user dictionary
+function ReaderUserHyph:openDictionary()
+    self.dict_file = self:getDictionaryPath()
+    self.new_dict_file = self.dict_file .. ".new"
+
+    -- if no file, crate an empty one
+    if lfs.attributes(self.dict_file, "mode") ~= "file" then
+        self.dict = io.open(self.dict_file, "w")
+        self.dict:close()
+    end
+
+    -- open files
+    self.dict = io.open(self.dict_file, "r")
+    if not self.dict then
+        logger.err("UserHyph: could not open " .. self.dict_file)
+        return
+    end
+    self.new_dict = io.open(self.new_dict_file, "w")
+    if not self.new_dict then
+        logger.err("UserHyph: could not open " .. self.new_dict_file)
+        return
+    end
+    return true
+end
+
+-- Reads the first dictionary entries from the old dictionary
+-- and writes them to the new dict, as long the entry is alphabetically lower than (_case insensitive_) word
+-- the last line is not written, but returned as: word, line
+--
+-- the format is line per line: "word;hyphenation\n"
+-- e.g.: "danger;dan-ger\n"
+function ReaderUserHyph:findEntry(word)
+    -- scan hyphenation dictionary for selected word
+    local word_lower = self.ui.document:getLowercasedWord(word)
+    local line = self.dict:read()
+    while line and self.ui.document:getLowercasedWord(line:sub(1, line:find(";") - 1)) < word_lower do
```

Comment:
Do you store only lowercase versions of word and its hyphenation in the file?
If yes, no need to call document:getLowercasedWord() on *each word* in the file :)
If no, could this be done?

Reply:
It could be done, to store words and hyphenation in lowercase, but there is no need for, as crengine converts for hyphenation everything to lowercase.
The highlighted word (from KOReader) is mixed case, so have to convert one word anyway.
It might seem awkward if you have Ferman nouns in lowercase stored. (And my friends form dante have their wordlist mixed-case too). So I would prefer this not to change.

Reply:
OK.
(Would just avoid lowercasing 10000 words each time you add one :)